File: trees/98. Validate Binary Search Tree/isValidBST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def checkBST(self, node, left, right) -> bool: 
        if not node:
            return True
        
        if not (node.val > left and node.val < right):
            return False
        
        return self.checkBST(node.left, left, node.val) and self.checkBST(node.right, node.val, right)
        
        
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.checkBST(root, float(-inf), float(inf)) 

File: trees/98. Validate Binary Search Tree/isValidBST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
var checkBST = (node, left, right) => {
    if (!node) {
        return true;
    }

    if (!(node.val > left && node.val < right)) {
        return false;
    }

    return checkBST(node.left, left, node.val) && checkBST(node.right, node.val, right);
}

/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
    return checkBST(root, -Infinity, Infinity)
};